### PR TITLE
skip mark mhlo.where as shape calc op

### PR DESCRIPTION
**Pull request by Orion34-lanbo**
This pr is prepared for `mhlo_disc.where` op's fusion implementation as mentioned in #742 

`mhlo.disc` op will produce 2 outputs, the 2nd output is i64 tensor with single element
represent number of valid elements in the 1st output, this will be used to calc the final output
shape used for `mhlo.real_dynamic_slice`.

 It is indeed a shape calculation, but where op should not be considered as shape calc op, it is more than that.
By marking it as shape calc op will cause `mhlo_disc.where` impossible to fuse in later fusion pass. Evenmore, the element-wise ops produces the input for `mhlo_disc.where` will be marked as shape calc ops, thus they also cannot be fused in later passes.

With changes from current pr, we can generate the following IR(no disc.shape_op attr for where op), which can be properly fused as we expected when we support fusion for  #`mhlo.where`.
```
module {
  func.func @main(%arg0: tensor<?x?x?xi64>) -> tensor<?x3xi64> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c2 = arith.constant 2 : index
    %c3 = arith.constant 3 : index
    %0 = mhlo.constant dense<0> : tensor<i64>
    %cst = arith.constant dense<1> : tensor<2xindex>
    %cst_0 = arith.constant dense<0> : tensor<2xindex>
    %dim = tensor.dim %arg0, %c0 : tensor<?x?x?xi64>
    %dim_1 = tensor.dim %arg0, %c1 : tensor<?x?x?xi64>
    %dim_2 = tensor.dim %arg0, %c2 : tensor<?x?x?xi64>
    %from_elements = tensor.from_elements %dim, %dim_1, %dim_2 {disc.shape_op = true} : tensor<3xindex>
    %1 = "mhlo.dynamic_broadcast_in_dim"(%0, %from_elements) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i64>, tensor<3xindex>) -> tensor<?x?x?xi64>
    %2 = mhlo.compare  NE, %arg0, %1 : (tensor<?x?x?xi64>, tensor<?x?x?xi64>) -> tensor<?x?x?xi1>
    %index, %num_output_elements = "mhlo_disc.where"(%2) : (tensor<?x?x?xi1>) -> (tensor<?x3xi64>, tensor<1xi64>)
    %extracted = tensor.extract %num_output_elements[%c0] : tensor<1xi64>
    %3 = arith.index_cast %extracted : i64 to index
    %from_elements_3 = tensor.from_elements %3, %c3 {disc.shape_op = true} : tensor<2xindex>
    %4 = mhlo.real_dynamic_slice %index, %cst_0, %from_elements_3, %cst : (tensor<?x3xi64>, tensor<2xindex>, tensor<2xindex>, tensor<2xindex>) -> tensor<?x3xi64>
    return %4 : tensor<?x3xi64>
  }
}
```

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc [hunk 1]
```diff
@@ -90,6 +90,19 @@ void DiscMarkShapeCalc::MarkShapeCalcOps() {
   }
 }
 
+// NOTE(lanbo.llb): mhlo.disc op will produce 2 outputs, the 2nd output is i64 tensor with single element
+// represent number of valid elements in the 1st output, this will be used to calc the final output
+// shape. It is indeed a shape calc, but where op should not be considered as shape calc op.
+// By marking it as shape calc op will cause mhlo_disc.where impossible to fuse in later fusion pass.
+bool with_input_from_where_op(Operation* op) {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc:97]: nit: Naming style...

Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc]: fixed

File: tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc [hunk 2]
```diff
@@ -90,6 +90,21 @@ void DiscMarkShapeCalc::MarkShapeCalcOps() {
...
 
+// NOTE(lanbo.llb): mhlo.disc op will produce 2 outputs, the 2nd output is i64
+// tensor with single element represent number of valid elements in the 1st
+// output, this will be used to calc the final output shape. It is indeed a
+// shape calc, but where op should not be considered as shape calc op. By
+// marking it as shape calc op will cause mhlo_disc.where impossible to fuse in
+// later fusion pass.
+bool withInputFromWhereOp(Operation* op) {
+  for (Value operand : op->getOperands()) {
+    if (isa<mhlo_disc::WhereOp>(operand.getDefiningOp())) {
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc:101]: If the operand is the first operand of where op, should we also need to treat it specially?

**Comment by qiuxiafei** [comment]
LGTM ... let @wyzero give the final LGTM.

File: tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc (hunk 2 above)
Comment (Orion34-lanbo) [reply tao_compiler/mlir/disc/transforms/mhlo_mark_shape_calc.cc]: Currently based on the computation pattern we have observed, we want to treat `mhlo_disc.where` as compute op, thus we do not treat them differently.

**Review by wyzero — APPROVED** [review]
LGTM

**Pull request merged**